drivers/aip31068/aip31068.c

```diff
@@ -0,0 +1,516 @@
+}
+
+int aip31068_set_text_insertion_mode(aip31068_t *dev,
+                                     aip31068_text_insertion_mode_t mode)
+{
+    if (mode == RIGHT_TO_LEFT) {
+        dev->_curr_entry_mode_set &= ~(1 << BIT_ENTRY_MODE_INCREMENT);
+    }
+    else {
+        dev->_curr_entry_mode_set |= (1 << BIT_ENTRY_MODE_INCREMENT);
+    }
+
+    return _command(dev, CMD_ENTRY_MODE_SET | dev->_curr_entry_mode_set);
+}
+
+int aip31068_move_cursor_left(aip31068_t *dev)
+{
+    uint8_t cmd = CMD_CURSOR_DISPLAY_SHIFT;
+    cmd &= ~(1 << BIT_CURSOR_DISPLAY_SHIFT_DIRECTION);
+
+    return _command(dev, cmd);
+}
+
+int aip31068_move_cursor_right(aip31068_t *dev)
+{
+    uint8_t cmd = CMD_CURSOR_DISPLAY_SHIFT;
+    cmd |= (1 << BIT_CURSOR_DISPLAY_SHIFT_DIRECTION);
+
+    return _command(dev, cmd);
+}
+
+int aip31068_scroll_display_left(aip31068_t *dev)
+{
+    uint8_t cmd = CMD_CURSOR_DISPLAY_SHIFT;
+    cmd |= (1 << BIT_CURSOR_DISPLAY_SHIFT_SELECTION);
+    cmd &= ~(1 << BIT_CURSOR_DISPLAY_SHIFT_DIRECTION);
+
+    return _command(dev, cmd);
+}
+
+int aip31068_scroll_display_right(aip31068_t *dev)
+{
+    uint8_t cmd = CMD_CURSOR_DISPLAY_SHIFT;
+    cmd |= (1 << BIT_CURSOR_DISPLAY_SHIFT_SELECTION);
+    cmd |= (1 << BIT_CURSOR_DISPLAY_SHIFT_DIRECTION);
+
+    return _command(dev, cmd);
+}
+
+int aip31068_set_custom_symbol(aip31068_t *dev,
+                               aip31068_custom_symbol_t custom_symbol,
+                               uint8_t charmap[])
+{
+    uint8_t location = custom_symbol;
+    int rc = _command(dev, CMD_SET_CGRAM_ADDR | (location << 3));
+
+    if (rc < 0) {
+        return rc;
+    }
+
+    int row_count = dev->params.font_size == FONT_SIZE_5x8 ? 8 : 10;
+
+    for (int i = 0; i < row_count; i++) {
+        rc = _data(dev, charmap[i]);
+
+        if (rc < 0) {
+            return rc;
+        }
+    }
+
+    // todo: return to old cursor position
+    return _command(dev, CMD_SET_DDRAM_ADDR);
+}
+
+int aip31068_print_custom_symbol(aip31068_t *dev,
+                                 aip31068_custom_symbol_t custom_symbol)
+{
+    return _data(dev, (uint8_t) custom_symbol);
+}
+
+int aip31068_print(aip31068_t *dev, const char *data)
+{
+    while (*data != '\0') {
+        int rc;
+        if ((rc = _data(dev, *data)) != 0) {
+            return rc;
+        }
+        data++;
+    }
+
+    return AIP31068_OK;
+}
+
+int aip31068_print_char(aip31068_t *dev, char c)
+{
+    return _data(dev, c);
+}
+
+void aip31068_set_progress_bar_enabled(aip31068_t *dev, bool enabled)
+{
+    dev->_progress_bar_enabled = enabled;
+
+    if (enabled) {
+        _init_progress_bar(dev, dev->params.row_count - 1);
+    }
+}
+
+void aip31068_set_progress_bar_row(aip31068_t *dev, uint8_t row)
+{
+    dev->_progress_bar_row = row;
+}
+
+void aip31068_set_progress(aip31068_t *dev, uint8_t progress)
+{
+    if (! dev->_progress_bar_enabled) {
+        return;
+    }
+
+    int bar_count = dev->params.col_count * 5;
```

Comment:
Magic number?

Reply:
Another name might be clearer, but basically it is the number of pixel-columns on a single line. `dev->params.col_count` is the number of characters for each line and that is taken times 5 because every characters has 5 pixel-columns.

Reply:
Maybe you can simply to a define for this: `CHAR_PIXEL_COLUMNS`? Could be better readability wise, probably also a `CHAR_PIXEL_ROWS` which would debed on the `FONT_SIZE`?